python/llm/src/bigdl/llm/transformers/speculative.py

```diff
@@ -596,7 +603,28 @@ def speculative_generate(self,
                     past_length = draft_past_key_values[0][0].size(2)
                     position_ids = torch.Tensor([[past_length]]).long().to(self.device)
                     forward_args["position_ids"] = position_ids
-                draft_output = draft_model(**forward_args)
+
+                if _enable_ipex:
+                    if "llama" in self.config.model_type:
+                        past_key_value_len = draft_past_key_values[0][0].shape[2]
+                        position_ids = torch.Tensor([[past_key_value_len + step_draft]]).long()
+                        # position_ids = draft_attention_mask.long().cumsum(-1) - 1
+                        position_ids = position_ids[:, :-draft_current_input_ids.size(0)]
+                        draft_output = draft_model.trace_graph(
+                            input_ids=draft_current_input_ids,
+                            attention_mask=draft_attention_mask,
+                            position_ids=position_ids,
+                            past_key_values=draft_past_key_values,
+                        )
+                    else:
+                        invalidInputError(False, "Only support Llama yet.")
+
+                    draft_output = CausalLMOutputWithPast(
+                        logits=draft_output[0],
+                        past_key_values=draft_output[1],
```

Comment:
How can one run speculative decoding for, say, ChatGLM3, using IPEX OPT now?

Reply:
Should be similar to llama, I'm verifying speculative decoding for some other models and will update soon.

Reply:
llama2, vicuna, chatglm3, baichuan2, qwen, mistral models have been verified on SPR.